core/domain/classifier_services.py

```diff
@@ -424,41 +375,240 @@ def delete_classifier_training_job(job_id):
+        state_training_jobs_mapping.algorithm_ids_to_job_ids.to_dict())
+    state_training_jobs_mapping_model.put()
+
+
+def get_classifier_training_job_maps(exp_id, exp_version, state_names):
+    """Gets the list of classifier training jobs mapped to algorithm id from
+    the exploration attributes.
 
     Args:
         exp_id: str. ID of the exploration.
         exp_version: int. The exploration version.
         state_names: list(str). The state names for which we retrieve the job.
 
     Returns:
-        list(ClassifierTrainingJob). Domain objects for the Classifier training
-        job model.
+        list(dict(str: ClassifierTrainingJob)). Domain objects for the
+        Classifier training job model for algorithm_ids.
     """
-    training_job_exploration_mapping_models = (
-        classifier_models.TrainingJobExplorationMappingModel.get_models(
+    state_training_jobs_mapping_models = (
+        classifier_models.StateTrainingJobsMappingModel.get_models(
             exp_id, exp_version, state_names))
-    job_ids = []
-    for mapping_model in training_job_exploration_mapping_models:
+    algorithm_ids_to_job_ids = []
+    for mapping_model in state_training_jobs_mapping_models:
         if mapping_model is None:
             continue
-        job_ids.append(mapping_model.job_id)
+        algorithm_ids_to_job_ids.append(
+            classifier_domain.AlgorithmIdToJobIdMapping(
+                mapping_model.algorithm_ids_to_job_ids))
+
+    job_id_to_algorithm_id_mapping = {
+        job_id: {
+            'algorithm_id': algorithm_id,
+            'state_name': mapping_model.state_name
+        } for mapping_model, map in python_utils.ZIP(
+            state_training_jobs_mapping_models,
+            algorithm_ids_to_job_ids)
+        for algorithm_id, job_id in map.algorithm_ids_to_job_ids.items()
+    }
+    job_ids = job_id_to_algorithm_id_mapping.keys()
 
     classifier_training_job_models = (
         classifier_models.ClassifierTrainingJobModel.get_multi(job_ids))
 
-    classifier_training_jobs = []
+    state_to_algorithm_id_job_id_map = {}
     for job_model in classifier_training_job_models:
-        classifier_training_jobs.append(get_classifier_training_job_from_model(
-            job_model))
-
-    # Backfill None's to maintain indexes.
-    for index, mapping_model in enumerate(
-            training_job_exploration_mapping_models):
-        if mapping_model is None:
-            classifier_training_jobs.insert(index, None)
-    return classifier_training_jobs
+        job_id = job_model.id
+        algorithm_id = job_id_to_algorithm_id_mapping[job_id]['algorithm_id']
+        state_name = job_id_to_algorithm_id_mapping[job_id]['state_name']
+        if state_name not in state_to_algorithm_id_job_id_map:
+            state_to_algorithm_id_job_id_map[state_name] = {}
+        state_to_algorithm_id_job_id_map[state_name][algorithm_id] = job_model
+
+    algorithm_id_to_training_job_maps = []
```

Comment:
OK -- I confess this is the point where I started to lose the plot completely for this function when reading it, and started trying to work it out on paper :-)

You already calculated algorithm_ids_to_job_ids above. Don't you just need to replace job_ids by the corresponding job_models to get the result you need? Why do we need lines 584-593 and 598-605 -- couldn't you just collect all the job IDs, get the job models, and do the replacement?

Reply:
You are right some reduction in code is possible, we can change `algorithm_ids_to_job_ids` to `state_to_algorithm_id_job_id_map` in the very first for loop itself. Then aggregate all from this map and match the job_ids with their corresponding job_model. That way 584-593 and 598-605 will both be reduced in number of lines of code but would still be required.

Initially, I didn't preserve the state information in `algorithm_ids_to_job_ids` list, which led to the bloating of code because now it had to be preserved.

Reply:
Why does the state info need to be preserved? As in, once you have the algorithm_ids_to_job_ids, why can't you replace job_id with job model (for each element of the list)?

Reply:
Done

Reply:
It needs to be preserved so that we can filter to non-None values (state for which mapping doesn't exist) at top and then insert them again at the bottom, otherwise we have to assume that a value may be None or it may be a dict. This can be done but former seems cleaner code.

Reply:
So all this stuff is just to ensure that elements that are None in the original list stay as None in the final list?

If so -- not sure why things need to be so complex in the middle in order to preserve that. It seems like something that could be easily checked for when constructing the final list (unless I'm misunderstanding).

Reply:
I think there is a bit misunderstanding here. Initially, `algorithm_ids_to_job_ids` contain only non-None values which makes it easier to just go through this list and replace job_ids with job_models without having to worry about a None entry in between. So later on, we go through state_names and insert Nones whenever we don't have the final mapping. Also, I have refactored this code based on your suggestions. So this code (above this comment) is not longer valid. I suggest we continue this discussion after I commit the changes, to see if the refactored code needs to be refactored again.

Reply:
That makes sense -- sgtm. Thanks!